packages/big-design/src/components/Tree/TreeNode/TreeNode.tsx

```diff
@@ -0,0 +1,303 @@
+import { CheckIcon, ChevronRightIcon, FolderIcon } from '@bigcommerce/big-design-icons';
+import React, { Dispatch, useCallback, useEffect, useMemo, useRef } from 'react';
+
+import { typedMemo } from '../../../utils';
+import { StyledCheckbox } from '../../Checkbox/private';
+import { FlexItem, FlexItemProps } from '../../Flex';
+import { StyledRadio } from '../../Radio/styled';
+import { Text } from '../../Typography';
+import { useIsExpanded, useIsSelected } from '../hooks';
+import { Action } from '../reducer';
+import { StyledUl } from '../styled';
+import { TreeNodeProps, TreeNodeRef, TreeProps, TreeState } from '../types';
+
+import { StyledArrowWrapper, StyledFlex, StyledGap, StyledLi, StyledSelectableWrapper } from './styled';
+
+interface PrivateTreeItemProps<T> {
+  state: TreeState<T>;
+  dispatch: Dispatch<Action<T>>;
+  iconless?: boolean;
+  selectable: TreeProps<T>['selectable'];
+  onExpand?: TreeProps<T>['onExpand'];
+  onCollapse?: TreeProps<T>['onCollapse'];
+  onSelect?: TreeProps<T>['onSelect'];
+}
+
+const flexItemProps: FlexItemProps = { flexShrink: 0, marginLeft: 'xxSmall' };
+
+const InternalTreeNode = <T extends unknown>({
+  children,
+  disabled,
+  dispatch,
+  icon,
+  iconless,
+  id,
+  label,
+  onCollapse,
+  onExpand,
+  onSelect,
+  selectable,
+  state,
+  value,
+}: TreeNodeProps<T> & PrivateTreeItemProps<T>): React.ReactElement<TreeNodeProps<T>> => {
+  const thisRef = useRef<TreeNodeRef<T>>({ children });
+  const nodeRef = useRef<HTMLLIElement | null>(null);
+  const selectableRef = useRef<HTMLLabelElement | null>(null);
+  const expanded = useIsExpanded(state, id);
+  const selected = useIsSelected(state, value);
+
+  // Could be multiple elements in which are clicked.
+  // Typing to generic Element type since all other elements extend from it.
+  const handleNodeToggle = async (e?: React.MouseEvent<Element>) => {
+    dispatch({ type: 'FOCUS', id });
+
+    // Prevents the collapse/expand when clicking on a radio or checkbox
+    // Checks to see if every element inside the selectableRef gets clicked.
+    if (e?.target instanceof Node && selectableRef.current?.contains(e?.target)) {
+      return;
+    }
+
+    if (thisRef.current.children) {
```

Comment:
:tropical_drink: to avoid nesting statements, can we return early here? Something like:

```js
if (thisRef.current.children.length === 0) {
  return;
}
```